fem/gridfunc.hpp

```diff
@@ -1561,10 +1561,176 @@ class GridFunction : public Vector
        must be 2 and that quad elements will be broken into two triangles.*/
    void SaveSTL(std::ostream &out, int TimesToRefine = 1);
 
+   /** @name Bounds of the gridfunction
+       \brief Piecewise linear bounds of bases can be used to compute bounds on the grid function in each element. The bounds for the bases are stored in \ref PLBound based on the following parameters:
+
+       (i) @b nb: number of bases/nodes in 1D (i.e. polynomial order+1),
+
+       (ii) @b b_type: bases type, 0 - Lagrange interpolants on Gauss-Legendre nodes, 1 - Lagrange interpolants on Gauss-Lobatto-Legendre nodes, and
+       2 - Positive/Bernstein bases on uniformly distributed nodes,
+
+       (iii) @b ncp: number of control points used to construct the
+       piecewise linear bounds
+
+       (iv) @b cp_type: control point distribution. 0 - GL + end-points,
+                        1 - Chebyshev.
+
+       Note: @b nb and @b b_type are infered directly from the grid-function.
+
+       If the user does not specify @b ncp and @b cp_type, the minimum value of
+       @b ncp is used that would bound the bases for the @b cp_type. We default
+       to @b cp_type = 0 as it requires fewer number of points to bound the bases. Typically, @b ncp = 2 @b nb is sufficient to get fairly compact bounds, and increasing @b ncp results in tighter bounds.
+
+       Finally, only tensor-product elements are currently supported.
+
+       For more technical details see:
+       Mittal et al., "General Field Evaluation in High-Order Meshes on GPUs" &
+       Dzanic et al., "A method for bounding high-order finite element
+       functions: Applications to mesh validity and bounds-preserving limiters".
+    */
+   ///@{
+   /// Struct to store piecewise linear bounds on the bases.
+   struct PLBound
```

Comment:
In terms of code organization, could `PLBound` exist outside of `GridFunction` as a separate class (potentially in a separate source file)?

And then, could the new member functions of `GridFunction` (at least some of them) be either free functions, or member functions of `PLBound`?

Reply:
Yeah I was also leaning towards that because it can help keep gridfunc.hpp clean when we enhance PLBound in future.